src/symbol/quads.js

```diff
@@ -99,93 +102,102 @@ export function getGlyphQuads(anchor: Anchor,
-        const bl = new Point(x1, y2);
-        const br = new Point(x2, y2);
-
-        if (rotateVerticalGlyph) {
-            // Vertical-supporting glyphs are laid out in 24x24 point boxes (1 square em)
-            // In horizontal orientation, the y values for glyphs are below the midline
-            // and we use a "yOffset" of -17 to pull them up to the middle.
-            // By rotating counter-clockwise around the point at the center of the left
-            // edge of a 24x24 layout box centered below the midline, we align the center
-            // of the glyphs with the horizontal midline, so the yOffset is no longer
-            // necessary, but we also pull the glyph to the left along the x axis.
-            // The y coordinate includes baseline yOffset, thus needs to be accounted
-            // for when glyph is rotated and translated.
-            const center = new Point(-halfAdvance, halfAdvance - shaping.yOffset);
-            const verticalRotation = -Math.PI / 2;
-
-            // xHalfWidhtOffsetcorrection is a difference between full-width and half-width
-            // advance, should be 0 for full-width glyphs and will pull up half-width glyphs.
-            const xHalfWidhtOffsetcorrection = ONE_EM / 2 - halfAdvance;
-            const xOffsetCorrection = new Point(5 - shaping.yOffset - xHalfWidhtOffsetcorrection, 0);
-            const verticalOffsetCorrection = new Point(...verticalizedLabelOffset);
-            tl._rotateAround(verticalRotation, center)._add(xOffsetCorrection)._add(verticalOffsetCorrection);
-            tr._rotateAround(verticalRotation, center)._add(xOffsetCorrection)._add(verticalOffsetCorrection);
-            bl._rotateAround(verticalRotation, center)._add(xOffsetCorrection)._add(verticalOffsetCorrection);
-            br._rotateAround(verticalRotation, center)._add(xOffsetCorrection)._add(verticalOffsetCorrection);
+    for (const line of shaping.positionedLines) {
+        for (const positionedGlyph of line.positionedGlyphs) {
+            // The rects have an additional buffer that is not included in their size.
+            const glyphPadding = 1.0;
+            let rectBuffer = GLYPH_PBF_BORDER + glyphPadding;
+            let isSDF = true;
+            let pixelRatio = 1.0;
+            let lineOffset = 0.0;
+
+            const rotateVerticalGlyph = (alongLine || allowVerticalPlacement) && positionedGlyph.vertical;
+            const halfAdvance = positionedGlyph.metrics.advance * positionedGlyph.scale / 2;
+
+            // Align images and scaled glyphs in the middle of a vertical line.
+            if (allowVerticalPlacement && shaping.verticalizable) {
+                const scaledGlyphOffset = (positionedGlyph.scale - 1) * ONE_EM;
+                const imageOffset = (ONE_EM - positionedGlyph.metrics.width * positionedGlyph.scale) / 2;
+                lineOffset = line.lineOffset / 2 - (positionedGlyph.imageName ? -imageOffset : scaledGlyphOffset);
```

Comment:
When label is shaped in horizontal orientation, a line could be offset if there scaled glyph, or tall image. In vertical orientation `line.lineOffset / 2` shifts Y to the middle of the vertical line and from that, glyph is shifted based on it's respective offset. For example, wide `gb-national-rail.london-overground.london-tfl-rail.london-underground` would be in the center as well as tiny `dot-9`

<img src="https://user-images.githubusercontent.com/1001009/68929570-448f8780-0795-11ea-9303-ce2e76282429.png" width=60%>

@ansis if you have any ideas of how this can be simplified, please tell.

Reply:
Thanks, that helps